docs/NitroRPC.md

```diff
@@ -0,0 +1,161 @@
+The Requests are signed by the initiator and the Response are countersigned.
+
+The use of a server side universal millisecond timestamp builds a temper proof history.
+
+## General definition
+
+Every message is formatted in a fixed JSON array (size is always 4 elements) as follow:
+
+`[request_id, method, params, ts]`
+
+**request_id:** is an unsigned integer unique identifier of the request, response must copy the same request_id
+**method**: remote method name to be invoked as string
+**params**: method params is a json array, it can be empty or contain flexible mix of json types.
+**ts**: an unsigned integer 64 bit representing server timestamp in millisecond
+
+## NitroRPC Request
+
+In this example, the client is calling a method named `"subtract"` with positional parameters:
+
+```json
+{
+  "req": [1001, "subtract", [42, 23], 1741344819012],
+  cid: "0x3ce9bc2e2053488bde06de257541e05fc58fd8c4f149cca44f1c702fcbdbde0aa"
+  out: ["Serialized array of Allocations"]
+  "sig": ["0xa0ad67f51cc73aee5b874ace9bc2e2053488bde06de257541e05fc58fd8c4f149cca44f1c702fcbdbde0aa09bcd24456f465e5c3002c011a3bc0f317df7777d2"]
+}
+```
+
+- req: rpc message payload `[request_id, method, params, ts]`
+- cid: ChannelId from channel creation
+- out: Outcome allocations can be optional
+- sig: payload client stateHash signature
+
+The millisecond timestamp was returned previously from the server, it is used as a height for proof of history
+
+## NitroRPC Response
+
+For a successful invocation, the server might respond like this:
+
+```json
+{
+  "res": [1001, "subtract", [19], 1741344819814],
+  cid: "0x3ce9bc2e2053488bde06de257541e05fc58fd8c4f149cca44f1c702fcbdbde0aa"
+  out: ["Serialized array of Allocations"]
+  "sig": ["0xd73268362b04516451ec52170f5c8ca189d35d9ac5e9041c156c9f0faf9aebd2891309e3b2b5d8788578ab3449c96f7aa81aefb25482b53f02bac42c65f806e5"]
+}
+```
+
+- res: rpc message payload `[request_id, method, params, ts]`
+- cid: ChannelId from channel creation
+- out: Outcome allocations can be optional
+- sig: payload server response stateHash signature
+
+**ts**: server response with latest timestamp
+
+## Go Types
+
+```go
+type RPCMessage struct {
+  RequestID uint64
+  Method    string
+  Params    []any
+  Result    []any
+  Timestamp uint64
+}
+```
+
+## Solidity types
+
+```solidity
+struct RPCMessage {
+    uint64 requestID;
+    string method;
+    bytes[] params;
+    bytes[] result;
+    uint64 timestamp;
+}
+
+struct RPCError {
+  uint32 code;
+  string message;
+}
+```
+
+### NitroRPC StateHash
+
+Requester and Responder must sign the RPCStateHash build in the following way:
+
+```solidity
+keccak256(
+  abi.encode(
+    reqId,      // Request unique Id act as a nounce
+    method,     // Method name
+    params,     // Client method params
+    result,     // Server response
+    timestamp,  // timestamp from the request
+  )
+);
+
+// Alternatively from RPCMessage type
+keccak256(
+  abi.encode(
+    rpcMessage
+  )
+)
+```
```

Comment:
This is wrong. 
Requester must sign:
```solidity
keccak256(
   abi.encode(
      reqId,
      method,
      params,
      timestamp,
      channelId,
      allocations
   )
)
```
 whereas Responder must sign the same struct with addition of `result`.

Reply:
I'm not fully understand this part. But protocol has one visible flow: parties don't exchange with allocation and signatures (only appdata). And signing appdata seems to have little effect on channel `close`, `reset` and `checkpoint`. So either there are should be protocol on top of it for exchanging States (appdata + allocations + sigs) or rpc protocol should exchange States in a first place... Oh, no the same problem we had with Nitro

Reply:
Some NitroSpec are outdated, we need to cherry pick our NitroRPC.sol in the other branch